### PR TITLE
  GitHub Issue NOAA-EMC/GSI#450 Using global and regional ensembles simultaneously for fv3-lam GSI EnVar (add l_both_fv3sar_gfs_ens option)

### LINKED ISSUE #450: Using global and regional ensembles simultaneously for fv3-lam GSI  EnVar
Up to now, in the EMC regional (FV3-LAM) DA parallel runs,  the pre-set global ensembles (80) are used as the sampling covariances in the GSI EnVAR.  This is mainly for the resources required for a self-cycled FV3-LAM regional ensembles.    With the emerging computational resources,  EMC and collaborators are developing self-cycled FV3-LAM (regional) ensembles based GSI EnVAR with smaller ensemble size (less than 40).  Hence, using global and regional ensembles simultaneously for fv3-lam GSI  EnVar is expected to be potentially helpful for increase the actual sampling ensemble size and extract larger scale information from the global ensembles .  
The GSI EnVar approach (all ensemble members are mapped to one regular ensemble grid) make implementation of such such mixed ensemble function rather straightforward.

### PR DESCRIPTION
**Pull request by TingLei-daprediction**
This PR is to add the mixed ensemble option for fv3-lam GSI EnVar.  https://github.com/NOAA-EMC/GSI/issues/450.
It passed the developer's own verification tests for this branch on hera.  On hera,  It failed in a few GSI 's own regression tests for the scalability problem, which I think could be attributed to some hera's system performance change and could be disregard now. I am re-running those regression tests and would update here.  
@MichaelLueken-NOAA  I did the rebase for this change and please let me know if it doesn't meet the GSI PR requirement.

### CODE REVIEW

File: src/gsi/cplr_get_fv3_regional_ensperts.f90
Comment (MichaelLueken) [inline src/gsi/cplr_get_fv3_regional_ensperts.f90]: Please replace .ne. with /= on line 110.

File: src/gsi/cplr_get_fv3_regional_ensperts.f90
Comment (MichaelLueken) [inline src/gsi/cplr_get_fv3_regional_ensperts.f90]: If this line won't be used in the future, please remove commented out line 250.  If it will be used, please remove cltorg from the comment.

File: src/gsi/cplr_get_fv3_regional_ensperts.f90
Comment (MichaelLueken) [inline src/gsi/cplr_get_fv3_regional_ensperts.f90]: If this line won't be used in the future, please remove commented out line 260. If it will be used, please remove cltorg from the comment.

File: src/gsi/cplr_get_fv3_regional_ensperts.f90
Comment (MichaelLueken) [inline src/gsi/cplr_get_fv3_regional_ensperts.f90]: If this line won't be used in the future, please remove commented out line 513. If it will be used, please remove cltorg from the comment.

File: src/gsi/cplr_get_fv3_regional_ensperts.f90
Comment (MichaelLueken) [inline src/gsi/cplr_get_fv3_regional_ensperts.f90]: If this line won't be used in the future, please remove commented out line 542. If it will be used, please remove cltorg from the comment.

File: src/gsi/cplr_get_fv3_regional_ensperts.f90
Comment (MichaelLueken) [inline src/gsi/cplr_get_fv3_regional_ensperts.f90]: If this line won't be used in the future, please remove commented out line 545. If it will be used, please remove cltorg from the comment.

File: src/gsi/gsimod.F90
Comment (MichaelLueken) [inline src/gsi/gsimod.F90]: Please add at least one more space to indentation on lines 1753 and 1754.

File: src/gsi/gsimod.F90
Comment (MichaelLueken) [inline src/gsi/gsimod.F90]: Please add at least one more space to indentation on lines 1756.

File: src/gsi/gsimod.F90
Comment (MichaelLueken) [inline src/gsi/gsimod.F90]: Please add at least one more space to indentation on lines 1758.

File: src/gsi/get_gefs_for_regional.f90
Comment (MichaelLueken) [inline src/gsi/get_gefs_for_regional.f90]: While compiling in debug mode, n_ens on line 44 was identified as unused.  Please remove unused variable.

File: src/gsi/cplr_get_fv3_regional_ensperts.f90
Comment (lbi2018) [inline src/gsi/cplr_get_fv3_regional_ensperts.f90]: Remove line 260 is unused.

File: src/gsi/cplr_get_fv3_regional_ensperts.f90
Comment (lbi2018) [inline src/gsi/cplr_get_fv3_regional_ensperts.f90]: Remove unused comment line

File: src/gsi/cplr_get_fv3_regional_ensperts.f90
Comment (lbi2018) [inline src/gsi/cplr_get_fv3_regional_ensperts.f90]: Remove line 542 and 545

File: src/gsi/get_gefs_for_regional.f90
Comment (lbi2018) [inline src/gsi/get_gefs_for_regional.f90]: Remove line 48

**Review by lbi2018 — CHANGES_REQUESTED** [review]

File: src/gsi/get_gefs_for_regional.f90
Comment (MichaelLueken) [inline src/gsi/get_gefs_for_regional.f90]: Should lines 235 and 236 remain?  If they should, please indent both of these lines two spaces (these lines should align with line 240).

**Review by MichaelLueken — CHANGES_REQUESTED** [review]
Hi @TingLei-daprediction.  I have completed my review.  The code was compiled using Intel and GNU compilers without issue.  While compiling in debug mode, one variable was identified as unused (n_ens on line 44 of get_gefs_for_regional.f90).  The regression tests were run for the Intel, GNU, and debug executables and all passed successfully.  There were a few issues I've noted in my review (@lbi2018 has also noted several of these in her review).  Before making any changes to the source code, please update your branch to the latest authoritative repository using:

- git remote add upstream https://github.com/NOAA-EMC/GSI.git
- git remote update
- git rebase upstream/develop

Please correct the conflicts in src/gsi/gsimod.F90 and src/gsi/hybrid_ensemble_parameter.f90.

- git rebase --continue

Please apply the changes from @lbi2018 and my reviews.

- git push origin feature/gsi_fv3lam_mix_ens --force

If you have any questions or encounter any issues, please let me know.

File: src/gsi/get_gefs_for_regional.f90
Comment (TingLei-daprediction) [reply src/gsi/get_gefs_for_regional.f90]: @MichaelLueken-NOAA  Thank you !  I will take care of  your and Li's questions/suggesitons when I come back to the office next Wenseday.   Have a good weekend!

**Comment by TingLei-daprediction** [comment]
@lbi2018 and @MichaelLueken-NOAA  I had finished the changes as requested and pushed them back to the repository. 
The GSI 's own regression regression tests are still running since yesterday.  However, my own tests demonstrated the current branch with the changes from EMC GSI gives the identical results compared with previous branch. It is reasonable since in those runs only conventional obs are used and the changes from EMC GSI are mainly to improve optimization in more difficult situations when satellite obs are used.   Hence, the current PR is regarded as verified.

**Comment by MichaelLueken** [comment]
Hi @TingLei-daprediction.  I see that you have merged the latest authoritative repository to your `feature/gsi_fv3lam_mix_ens` branch.  However, it doesn't look as though you addressed the coding concerns brought up by @lbi2018 and myself.  Please make sure to go over our requested changes, make the necessary changes, use `git add` on the modified files, then `git commit --amend` to commit the changes, then `git push origin feature/gsi_fv3lam_mix_ens --force` to push the modified files back to the repository.  Thank you very much for your time!

**Comment by TingLei-daprediction** [comment]
@MichaelLueken-NOAA ,  My bad.  the local changes were not  committed back.  Now it should be ok.  Thanks.

File: src/gsi/get_gefs_for_regional.f90
Comment (MichaelLueken) [reply src/gsi/get_gefs_for_regional.f90]: @TingLei-daprediction Should the two writes here be kept?  They almost look like they are debug writes.  If they should remain, please indent these two lines by two spaces so that they proper align with line 238 below.

**Review by MichaelLueken — CHANGES_REQUESTED** [review]
@TingLei-daprediction Thank you very much for addressing @lbi2018 and my concerns!  On lines 234 and 235 of `src/gsi/get_gefs_for_regional.f90`, if these lines should be kept (they appear to be debug writes, at first glance), then please indent these two lines two spaces so that they align with line 238 below.  Once completed and once @lbi2018 has given her approval, I will be able to submit your work to the review committee.  Thanks again!

**Review by MichaelLueken — APPROVED** [review]
Thanks, @TingLei-daprediction!  I approve of these changes.

**Comment by TingLei-daprediction** [comment]
@MichaelLueken-NOAA  Thanks for pointing me to this line, around which I should give more careful treatment with more informative warning as the current version.

**Comment by MichaelLueken** [comment]
Hi @lbi2018.  At your earliest convenience, please go through the changes again and make sure that you are happy with the changes.  If you are, please approve the changes, otherwise, please let @TingLei-daprediction know.  Thank you very much!

**Review by lbi2018 — APPROVED** [review]

**Comment by lbi2018** [comment]
Approved the changes. Thanks. -Li

**Comment by MichaelLueken** [comment]
The due date for these changes have past with no feedback, so I will give final approval to these changes and merge them to the authoritative develop branch.

**Pull request merged**
